### PR TITLE
netctl cli autocompletion - static and dynamic

### PR DESCRIPTION
**Pull request by vijayvikrant**
(1) Added the Bash file for auto-completion. Covers about 70% cases - both static and dynamic completions
(2) Covers dynamic completions for tenants, policies, network names, groups 
(3) vlan/vxlan dynamic auto-completions are pending and will be added to the same PR soon.

##### tenant
```
vagrant@netplugin-node1:$ netctl
group    help     network  policy   rule     tenant
vagrant@netplugin-node1:$ netctl tenant
create  help    ls      rm
vagrant@netplugin-node1:$ netctl tenant rm
default  pepsi
```

##### network
```
vagrant@netplugin-node1:$ netctl network create --tenant
default  pepsi
vagrant@netplugin-node1:$ netctl network create --tenant
default  pepsi
vagrant@netplugin-node1:$ netctl network create --tenant pepsi --encap v
vlan   vxlan
vagrant@netplugin-node1:$ netctl network create --tenant pepsi --encap v
vlan   vxlan
vagrant@netplugin-node1:$ netctl network create --tenant pepsi --encap vxlan --
--encap    --gateway  --pkt-tag  --subnet   --tenant
vagrant@netplugin-node1:$
```

##### policy
```
vagrant@netplugin-node1:$ netctl policy
create  help    ls      rm
vagrant@netplugin-node1:$ netctl policy create -
-t        --tenant
vagrant@netplugin-node1:$ netctl policy create --tenant
default  pepsi
vagrant@netplugin-node1:$
vagrant@netplugin-node1:$ netctl policy rm [TAB]
vagrant@netplugin-node1:$ netctl policy rm prod_[TAB]
prod_db   prod_web
vagrant@netplugin-node1:$ netctl policy rm —tenant[TAB]
default  pepsi
vagrant@netplugin-node1:$ netctl policy rm --tenant pepsi [TAB]
vagrant@netplugin-node1:$ netctl policy rm --tenant pepsi prod_db1
vagrant@netplugin-node1:$
vagrant@netplugin-node1:$ netctl policy ls -
-a        --all     -j        --json    -t        --tenant
vagrant@netplugin-node1:$ netctl policy ls --tenant
default  pepsi
vagrant@netplugin-node1:$
```

##### groups
```
vagrant@netplugin-node1:$ netctl group
create  help    ls      rm
vagrant@netplugin-node1:$ netctl group create -
-p        --policy  -t        --tenant
vagrant@netplugin-node1:$
vagrant@netplugin-node1:$ netctl group create --tenant
default  pepsi
vagrant@netplugin-node1:$

vagrant@netplugin-node1:$ netctl group ls -
-a        --all     -j        --json    -t        --tenant
vagrant@netplugin-node1:$ netctl group ls --all
Tenant   Group  Network     Policies
------   -----  -------     --------
pepsi    web    pepsi-net1  prod_db1
default  web    pepsi-net   prod_web
vagrant@netplugin-node1:$ netctl group ls --tenant
default  pepsi
vagrant@netplugin-node1:$ netctl group ls --tenant pepsi
Tenant  Group  Network     Policies
------  -----  -------     --------
pepsi   web    pepsi-net1  prod_db1
vagrant@netplugin-node1:$

vagrant@netplugin-node1:$ netctl group rm [TAB]
vagrant@netplugin-node1:$ netctl group rm web

vagrant@netplugin-node1:$ netctl group rm —tenant [TAB]
default  pepsi
vagrant@netplugin-node1:$ netctl group rm --tenant pepsi [TAB]
vagrant@netplugin-node1:$ netctl group rm --tenant pepsi web
```

##### rule
```
vagrant@netplugin-node1:$ netctl rule
add   help  ls    rm
vagrant@netplugin-node1:$ netctl rule add -
-a            -d            -g            --ip-address  -l            --network     -P            --priority    -t
--action      --direction   --group       -j            -n            -p            --port        --protocol    --tenant
vagrant@netplugin-node1:$ netctl rule add --action
accept  deny
vagrant@netplugin-node1:$ netctl rule add --action accept --direction
both  in    out
vagrant@netplugin-node1:$ netctl rule add --action accept --direction both --tenant
default  pepsi
vagrant@netplugin-node1:$ netctl rule add --action accept --direction both --tenant default --network
coke-net   pepsi-net
vagrant@netplugin-node1:$ netctl rule add --action accept --direction both --tenant default --network coke-net

vagrant@netplugin-node1:$ netctl rule rm --tenant
default  pepsi
vagrant@netplugin-node1:$

vagrant@netplugin-node1:$ netctl rule ls --
--all     --json    --tenant
vagrant@netplugin-node1:$ netctl rule ls --tenant
default  pepsi
vagrant@netplugin-node1:$
```

### CODE REVIEW

File: contrib/completion/bash/netctl [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+
+GLOBAL_OPTIONS="\
+    --netmaster\
+    --help\
+    --version"
+
+# Fetch all tenant names
+_netctl_complete_tenants() {
+    fields='$1'
+    COMPREPLY=( $(compgen -W '$(netctl tenant ls | awk "NR>2 {print $fields}")' -- "$cur") )
```
Comment (jainvipin) [inline contrib/completion/bash/netctl]: @vijayvikrant - I thought you were going to add `-q` for various objects to avoid having to do awk, etc. i.e. `netctl tenant ls -q` would list the tenants which can then be used not only by you here, but also by the end user to automate things to enlist all object names to iterate upon.

Same comment for various other things.

**Comment by vijayvikrant** [comment]
@jainvipin we discussed about doing a --json earlier, but as @erikh said json parsing in bash would be painful.

**Comment by jainvipin** [comment]
> @jainvipin we discussed about doing a --json earlier, but as @erikh said json parsing in bash would be painful.

Agree doing --json is not a good idea. I thought we agreed to doing `-q` option for example `netctl tenant ls -q` that only lists the tenant names, and so you can use them instead of awking (because awking depends on screen formatting). This is along the lines of `docker ps -q` for example...

If we implement -q version it can be used not only by bash completion, but also by other users to use this feature in their automation/scripts/etc

**Comment by vijayvikrant** [comment]
@jainvipin.. Will do. Does the diff look good otherwise?

File: contrib/completion/bash/netctl [hunk 2]
```diff
@@ -0,0 +1,463 @@
+#! /bin/bash
+
+GLOBAL_COMMANDS="\
```
Comment (jainvipin) [inline contrib/completion/bash/netctl]: Must we specify these options explicitly, can we get it from reading things themselves? Same comment for GLOBAL_OPTIONS. I am worried about maintaining this file along with any change as we go forward. I hopt we don't add too many global vars and options, but having to specify at two places doesn't seem good.

Comment (jainvipin) [inline contrib/completion/bash/netctl]: Looking more at the rest of the code/logic, it does seem that we are dependent on the top level objects, etc quite a bit. I am not sure if generating this code (or at least the parts of the code) is worth the gain, because then we can truly become independent of future changes.
So my suggestion is to not move to auto-generation of it but exploring this may benefit us in long run with highly sustainable code. ping @shaleman

**Comment by vijayvikrant** [comment]
@jainvipin Made the changes for supporting -q flag for all ls commands, and changed the auto_complete file to use it. PTAL

**Comment by jainvipin** [comment]
`-q` options looks good. Let's take auto-generation of some of the bash completion code outside this PR.
Before LGTM, I want to make sure that we resolve the `policy` when adding a rule. Can you please make the changes for making it with `-policy` so if @shaleman, @mapuri and others are okay, you are ready to go with the checkin.

**Comment by vijayvikrant** [comment]
@jainvipin added the logic for rule and policy completions..

##### because policy is a mandatory flag, unless policy is specified the auto-completion will stop only the flags and not rules
```
vagrant@netplugin-node1:$ netctl rule rm -
-o        --policy  -t        --tenant
vagrant@netplugin-node1:$ netctl rule rm --tenant
coke     default  pepsi
vagrant@netplugin-node1:$ netctl rule rm --tenant pepsi -
-o        --policy  -t        --tenant
```

##### if a tenant is specified, then the policy completions are based of that tenant
```
vagrant@netplugin-node1:$ netctl rule rm --tenant pepsi --policy db_pepsi_prod
```

##### once the policy is specified, then the auto-completions will show the rules that apply to that policy
```
vagrant@netplugin-node1:$ netctl rule rm --policy
db_prod   web_prod
vagrant@netplugin-node1:$ netctl rule rm --policy web_prod
1  2  3
vagrant@netplugin-node1:$ netctl rule rm --policy web_prod 1
```

File: Makefile [hunk 3]
```diff
@@ -39,7 +39,8 @@ checks:
 run-build: deps checks clean
 	cd Godeps/_workspace/src/github.com/contiv/ && chmod -R 771 contivmodel/ && cd contivmodel/ && ./generate.sh && \
 	cd /opt/gopath/src/github.com/contiv/netplugin && \
-	godep go install -v $(TO_BUILD) 
+	godep go install -v $(TO_BUILD) && \
+	sudo cp contrib/completion/bash/netctl /etc/bash_completion.d/netctl
```
Comment (shaleman) [inline Makefile]: You might want to make the bash completion script part of the netplugin release tar file. Please see "make tar" target. That way, you can use the file in ansible scripts to setup bash completion.

You can do this as a separate PR if you wish.

Comment (erikh) [inline Makefile]: +1 to that, that way volplugin gets it too.

File: netctl/commands.go [hunk 4]
```diff
@@ -168,27 +174,37 @@ var Commands = []cli.Command{
...
+				Action: listRules,
 			},
 			{
 				Name:      "rm",
 				Aliases:   []string{"delete"},
 				Usage:     "Delete an existing rule.",
-				ArgsUsage: "[policy] [rule id]",
-				Flags:     []cli.Flag{tenantFlag},
-				Action:    deleteRule,
+				ArgsUsage: "[rule id]",
```
Comment (shaleman) [inline netctl/commands.go]: @vijayvikrant @jainvipin 
Bringing up the broader topic of policy and rule UX here. Just because policy and rules are different REST objects, user doesn't have to see it that way. We can present the rules as a sub-object of policy. If we had a UI, it would display a table of rules inside the policy object(In fact this is what our rudimentary web-ui does).

Given that, I think the right way to present this to the user is:
netctl policy create <policy-name>
netctl policy add-rule <policy-name> <rule-id> [<rule-options>]
netctl policy delete-rule <policy-name> <rule-id> [<rule-options>]
netctl policy list-rule <policy-name> OR netctl policy inspect <policy-name>
netctl policy rm <policy-name>

I based add-rule and delete-rule options here similar to "docker network connect" and "docker network disconnect" CLI. This will be more easier for people to visualize rules and policies.

Comment (shaleman) [inline netctl/commands.go]: @vijayvikrant on a side note, any changes to netctl CLI will require corresponding documentation changes too

File: Makefile (hunk 3 above)
Comment (jainvipin) [inline Makefile]: +1 on making this part of the release tar; because that would be needed when we make the changes for ansible repo to auto-configure this.

**Comment by shaleman** [comment]
LGTM other than the rule CLI change

File: Makefile (hunk 3 above)
Comment (vijayvikrant) [inline Makefile]: sure, will add this to the "tar"  target

File: netctl/commands.go (hunk 4 above)
Comment (vijayvikrant) [inline netctl/commands.go]: @shaleman While everything that you said makes sense and I agree with it. But, from a UX perspective, even if we keep auto-completion aside for a moment; there is still a slight element of ambiguity/extra-responsibility that we enforce on the user by making the positioning mandatory. `netctl policy add-rule <policy-name> <rule-id>` the policy-name and rule-id should be entered in the exact same order. This is not very intuitive as the user will have to see help to get to it. so, In my opinion, either we should do `netctl rule add --policy <policy-name> <rule-id>` or we should do `netctl policy add-rule --policy <policy-name> <rule-id>`. That way, the user can hit tabs and get to what he would want.

Comment (shaleman) [inline netctl/commands.go]: @vijayvikrant 
I can see that `netctl policy <policy-name> add-rule <rule-id> [rule-options]` might have been a better option. I was just basing it on `docker network connect` which takes two arguments in this form:
```
$ docker network connect --help

Usage:	docker network connect [OPTIONS] NETWORK CONTAINER

Connects a container to a network

  --help=false       Print usage
```
There are other example too like `docker run` takes an image-name and a command argument. We can follow those example i think.

Comment (vijayvikrant) [inline netctl/commands.go]: @shaleman ok, then can i will make the changes for it to be `netctl policy add-rule <policy-name> <rule-id>` `netctl policy rm-rule <policy-name> <rule-id>` `netctl policy ls-rule <policy-name>`
I will also make changes to auto-completion to try and mind the order by added some positioning logic in the auto-completion.

**Comment by vijayvikrant** [comment]
@shaleman I have made the changes as per the previous comment. PTAL, if and when you are ok with them - i will do the related changes in documentation..

this is how the autocomplete works for rule commands,
```
 vagrant@netplugin-node1:/opt/gopath/src/github.com/contiv/netplugin$ netctl policy rule-rm
db_prod   --tenant  web_prod
vagrant@netplugin-node1:/opt/gopath/src/github.com/contiv/netplugin$ netctl policy rule-rm web_prod
1  2  3
vagrant@netplugin-node1:/opt/gopath/src/github.com/contiv/netplugin$ netctl policy rule-rm web_prod 1
vagrant@netplugin-node1:/opt/gopath/src/github.com/contiv/netplugin$

vagrant@netplugin-node1:/opt/gopath/src/github.com/contiv/netplugin$ netctl policy rule-rm --tenant
coke     default  pepsi
vagrant@netplugin-node1:/opt/gopath/src/github.com/contiv/netplugin$ netctl policy rule-rm --tenant pepsi db_pepsi_prod
1  2
vagrant@netplugin-node1:/opt/gopath/src/github.com/contiv/netplugin$ netctl policy rule-rm --tenant pepsi db_pepsi_prod
1  2
vagrant@netplugin-node1:/opt/gopath/src/github.com/contiv/netplugin$
```

**Comment by shaleman** [comment]
LGTM
@vijayvikrant Thanks for making the change.
Please rebase and we can merge after that.

**Comment by vijayvikrant** [comment]
@shaleman I have rebased the branch to master.
